### ISSUE TITLE
Should no-extra-semi trigger for es6 class methods?

### ISSUE THREAD

**Issue by scottohara** [eslint/eslint#2794]
###### Disclaimers:
 1. I've only just started my ES6 journey (so this could be my misinterpretation of `class` semantics)
 2. I'm using the [babel-eslint](https://github.com/babel/babel-eslint) parser, if that matters (although the same behaviour seems to occur with the default parser)

###### Issue:
Given the class:
```
/*eslint semi: [2, "always"], no-extra-semi: 2*/

class Foo {
  constructor() {
    ...
  }         // <- note: no semicolon after closing brace

  bar() {
    ...
  };        // <- note: with semicolon after closing brace

  baz() {
    ...
  }         // <- note: no semicolon after closing brace
}
```

... no warnings are raised.  Should they be?

It would seem that semicolons are allowed, but optional, for ES6 class methods.

If so, I would have expected that either:
 1. the `semi` rule would complain about the missing semicolons on the `constructor()` and `baz()` methods; **or**
 2. the `no-extra-semi` rule would complain about the explicit semicolon after the `bar()` method?

**Comment by btmills** [comment]
Here's the relevant portion of the [spec](http://www.ecma-international.org/ecma-262/6.0/#sec-functions-and-classes):

![screenshot 2015-06-19 09 52 21](https://cloud.githubusercontent.com/assets/1709537/8254739/e2917eb6-1668-11e5-8779-13cd0c844b81.png)

So semicolons are allowed, but [per ESTree](http://www.ecma-international.org/ecma-262/6.0/#sec-functions-and-classes), they get dropped from the AST. I agree ESLint should catch that somehow, but I'm not sure as to the best way to address it. I'm happy to file an issue with ESTree if that's the best option, though it's probably more of a concern for the concrete syntax tree. Should we do something ourselves with the tokens instead?

**Comment by mysticatea** [comment]
`MethodDefinition` seems to not have semicolon.
Semicolon seems to make an empty `ClassElement`. 
It resembles that a semicolon makes an `EmptyStatement`.

So I think it's better that the `no-extra-semi` rule warns the semicolon after the `bar()`.

**Comment by nzakas** [comment]
Agree, we should warn on this. The semicolon is unnecessary.

**Comment by btmills** [comment]
Should the semicolon be included in the AST somehow, or will we need to check for semicolon tokens ourselves?

**Comment by nzakas** [comment]
Ooh you know, I missed your spec question. I think it's worth having an ESTree discussion about it.

**Comment by btmills** [comment]
See estree/estree#93

**Comment by btmills** [comment]
Unsurprisingly, that extra semicolon belongs in a CST but not an AST. That leaves two options: we could wait for estree/estree#41 or check manually for unnecessary tokens within class bodies.

**Comment by nzakas** [comment]
Let's just check manually. Rationally, it's no different than what we already do for function declarations.

**Comment by mysticatea** [comment]
I will try fixing this.

**Comment by andriibuts** [comment]
I'm not sure whether it is a good place to put this but I it would also be great if `semi` and `no-extra-semi` rule would work with class fields from this proposal - https://github.com/tc39/proposal-class-public-fields
